vector-im/element-android
# #5005: Add save media icon in gallery

## Type of change


## Content


A download icon has been added into the media gallery screen to be able to download the media (not only images) from that screen. Message actions "share", "save" and "remove" have been re-order in the message actions menu which appears on long press on message. 
Introducing UseCase pattern to better encapsulate code and to be able to unit test (see interface `VectorBaseInOutUseCase`).

## Motivation and context


Closes #5005.

## Screenshots / GIFs


<img src="https://user-images.githubusercontent.com/46314705/155683422-49c7469b-3f7e-4fb0-9c84-d0708e16feb5.png" width=300/><img src="https://user-images.githubusercontent.com/46314705/155683433-6fab7c40-4e7b-4cd9-9a09-9f25de4273ee.png" width=300/>
<img src="https://user-images.githubusercontent.com/46314705/155683445-75517bc4-e006-46c9-8bb2-4606f93d596e.png" width=300/>

## Tests



- Go to a room where there are shared media (e.g. image, video)
- Long press on a shared media
- Check the order of actions in the appearing menu is correct (see [this comment](https://github.com/vector-im/element-android/issues/5005#issuecomment-1020463094))
- Cancel the menu
- Click on a shared media
- Check the media is displayed in the gallery
- Check download icon is shown along share icon
- Click on the download icon
- Check if write permissions is asked (only for devices with version < Android 10)
- Check a notification appears saying the media has been downloaded

## Tested devices

- OS version(s): Android 9 and 11

## Checklist
